Gracefully handle DO lets_encrypt cert rotation

waynr opened the pull request:
While DOKS does not officially support Let's Encrypt, it is possible for users to create a DO Let's Encrypt certificate and use that cert UUID for the "service.beta.kubernetes.io/do-loadbalancer-certificate-id" on Services they create in their cluster.

This is problematic because some combination of the Certificates Service and LBaaS will automatically rotate Let's Encrypt certificates which deletes the old Certificate (resulting an invalid UUID on the "service.beta.kubernetes.io/do-loadbalancer-certificate-id" Service annotation) then updates the Loadbalancer to point at a new certificate with CCM none the wiser.

Because of the following facts:

* CCM doesn't "know" about Let's Encrypt certificates
* CCM always attempts to set the cert UUID from the Service on the LoadBalancer when making any kind of update to it
* LBaaS will return HTTP 404 status code when it determines that there is no Certificate pointed to by the given cert UUID

CCM subsequently cannot make any change to LoadBalancers in these scenarios. Actually, this can happen when users specify their own custom Certificate and update a LoadBalancer manually without also updating the Service object, but that is probably more rare than the Let's Encrypt scenario and is arguably the user's responsibility not to shoot themselves in the foot by manual changing a LoadBalancer managed by CCM.

This PR proposes a fix that changes the way CCM behaves when it encounters a Let's Encrypt certificate associated with the LoadBalancer that is different from the one found on the Service annotation; specifically, it makes an effort to update the Service annotation with the correct cert UUID.

Left to do before merging this PR:

* [x] Update broken test cases
* [x] add new test cases to cover the different cert type and existence scenarios
* [x] Consider how to handle the known edge case where the load balancer has a valid LE cert but the user may have specified a new certificate ID manually
  * Internal discussion within the DOKS team has initially proposed that we don't support this use case, but we should consider documenting that a user can't (easily) switch from a lets_encrypt to a user-managed certificate.
  * Use certificate creation datetime to determine precedence when the LB and the service certificate differe and both are valid certs.
* [x] collapse service patching code into single call
* [x] create separate PR for service cert id missing 
* [x] use loadbalancer request creation for validation (use in-memory validation where possible before reaching out to APIs)
* [x] if we can't find the loadbalancer certificate we should not fail immediately but should still attempt to update from service (add test case to validate the correct behavior here)
* [ ] manually create service/lb using k8saas-images PR image

Refs #92

snormore commented on cloud-controller-manager/do/loadbalancers.go:
```diff
@@ -275,6 +274,57 @@ func (l *loadBalancers) EnsureLoadBalancer(ctx context.Context, clusterName stri
+		}
+
+		if lbCert.Type == "lets_encrypt" {
+			l.ensureCertificateIDAnnot(service, lbCertID)
```
Should this set a the new `serviceCertID = getCertificateID(service)` here too?

waynr replied:
It's not strictly necessary; at the line just below this we set `checkServiceCertID` to `false` and the block conditioned on that variable is the only place where the service cert id is used.

That being said, I can see how someone else coming at this code later on might be confused so it probably does make sense to set it to over write the old value with that of `lbCertID`.

waynr replied:
Alternatively, it `getCertificateID` is a fairly cheap call that just indexes into a map to return a value...while I know next to nothing about the go compiler or compilers in general, i wouldn't be surprised to learn functions like that are simply inlined... [the doc on go compiler optimizations suggests this is probably the case](https://github.com/golang/go/wiki/CompilerOptimizations#function-inlining)

So where I'm going with this is, I think I'll just replace all `serviceCertID` instances with the function call and avoid the potential for confusion altogether.

timoreimann replied:
The extra lookup is certainly not going to affect us performance-wise.

snormore commented on cloud-controller-manager/do/loadbalancers.go:
```diff
@@ -275,6 +274,57 @@ func (l *loadBalancers) EnsureLoadBalancer(ctx context.Context, clusterName stri
+			return fmt.Errorf("failed to get current certificate for lb")
+		}
+
+		if lbCert.Type == "lets_encrypt" {
```
Pull this out into a const?

snormore commented on cloud-controller-manager/do/loadbalancers.go:
```diff
@@ -275,6 +274,57 @@ func (l *loadBalancers) EnsureLoadBalancer(ctx context.Context, clusterName stri
+		_, _, err := l.resources.gclient.Certificates.Get(ctx, serviceCertID)
+		if err != nil {
+			respErr, ok := err.(*godo.ErrorResponse)
+			if ok && respErr.Response.StatusCode == 404 {
```
s/`404`/`http.StatusNotFound`

timoreimann commented on cloud-controller-manager/do/loadbalancers.go:
```diff
@@ -275,6 +274,56 @@ func (l *loadBalancers) EnsureLoadBalancer(ctx context.Context, clusterName stri
+			}
+			return nil, err
+		}
+	}
```
I think the logic around certificate validation should be moved into a dedicated function. AFAICS, that would also allow us to replace the `checkServiceCertID` variable / check by a premature return.

waynr replied:
My inclination is to factor out functions in the following situations:

* there is an immediate need to reuse a snippet of code
* there will obivously be need to reuse a snippet of code
* the function where the snippet currently resides is getting too unwieldy

I don't think any of those criteria apply in this situation, but I'll take a look once I've had an opportunity to add some test cases to validate what is here so far.

timoreimann replied:
Code fluency and natural scoping of functionality are additional concerns to me. Right now, `updateLoadBalancer` consists of "do stuff around certificates" and the actual update. The former isn't obvious on first sight, especially since it's not core business to updating.

I also think the code would be easier to understand as we could get rid of `checkServiceCertID`: No need to set and check conditionals when a `return nil` has the same effect.

timoreimann commented on cloud-controller-manager/do/loadbalancers.go:
```diff
@@ -275,6 +274,56 @@ func (l *loadBalancers) EnsureLoadBalancer(ctx context.Context, clusterName stri
+			respErr, ok := err.(*godo.ErrorResponse)
+			if ok && respErr.Response.StatusCode == http.StatusNotFound {
+				fmtStr := "certificate UUID %q not found; the %q service annotation refers to nonexistent DO Certificate"
+				respErr.Message = fmt.Sprintf(fmtStr, getCertificateID(service), annDOCertificateID)
```
I think we shouldn't overwrite the error message from an error we do not own. Creating a new error seems like the better option to me.

Doing so would also allow us to replace the type assertion by checking the response directly like

```
if resp != nil && resp.StatusCode == http.StatusNotFound {
  return fmt.Errorf(...)
}
```

(This would replace line 310 forward.)

timoreimann commented on cloud-controller-manager/do/certificates.go:
```diff
@@ -0,0 +1,7 @@
+package do
```
You need to add the licence header to this file as well.

waynr replied:
doh, thanks for the reminder

timoreimann commented on cloud-controller-manager/do/loadbalancers.go:
```diff
@@ -290,12 +339,9 @@ func (l *loadBalancers) UpdateLoadBalancer(ctx context.Context, clusterName stri
 
-	_, _, err = l.resources.gclient.LoadBalancers.Update(ctx, lb.ID, lbRequest)
-	if err != nil {
-		return fmt.Errorf("failed to update load-balancer with ID %s: %s", lb.ID, err)
```
The intention here is to annotate the error because the upstream error itself does not provide enough context (at least it didn't last time I checked).

waynr replied:
I replaced the `l.resources.gclient.LoadBalancers.Update` and subsequent error check with a call to `l.getLoadBalancer` which [returns an identically-formatted error message when the update fails](https://github.com/digitalocean/digitalocean-cloud-controller-manager/pull/274/files#diff-0f171474489b3a27add24a94c0da1505R321)

timoreimann commented on cloud-controller-manager/do/loadbalancers_test.go:
```diff
@@ -3411,13 +3411,24 @@ func Test_GetLoadBalancer(t *testing.T) {
 		listFn            func(context.Context, *godo.ListOptions) ([]godo.LoadBalancer, *godo.Response, error)
 		createFn          func(context.Context, *godo.LoadBalancerRequest) (*godo.LoadBalancer, *godo.Response, error)
 		updateFn          func(ctx context.Context, lbID string, lbr *godo.LoadBalancerRequest) (*godo.LoadBalancer, *godo.Response, error)
+		getCertFn         func(context.Context, string) (*godo.Certificate, *godo.Response, error)
```
We never seem to set this in any test case?

waynr replied:
This is only intended to keep the existing tests from breaking with the addition of `kvCertService` to the `newFakeClient` signature.

timoreimann commented on cloud-controller-manager/do/loadbalancers.go:
```diff
@@ -275,6 +274,56 @@ func (l *loadBalancers) EnsureLoadBalancer(ctx context.Context, clusterName stri
+	if lbCertID != "" && lbCertID != getCertificateID(service) {
+		lbCert, _, err := l.resources.gclient.Certificates.Get(ctx, lbCertID)
+		if err != nil {
+			return nil, fmt.Errorf("failed to get current certificate for lb")
```
This one should include the `err`.

waynr commented on cloud-controller-manager/do/loadbalancers.go:
```diff
@@ -275,6 +274,57 @@ func (l *loadBalancers) EnsureLoadBalancer(ctx context.Context, clusterName stri
+			// on the same loadbalancer to have different certificates associated
+			// with them -- as far as we are concerned where a load balancer is
+			// created by CCM in the first place it doesn't seem possible so we only
+			// need to grab the first one we find
```
> The LB API supports different certificates in different forwarding rules;

Yeah, my comment is intended to explain why we are only paying attention to the first one found.

> terms of what is possible

So just to be clear, you want me to clarify in my comment (which already frames it in terms of what is possible for a CCM-managed LB) that CCM support for multiple TLS serving certificates may be possible in the future?

waynr commented on cloud-controller-manager/do/loadbalancers.go:
```diff
@@ -275,6 +274,57 @@ func (l *loadBalancers) EnsureLoadBalancer(ctx context.Context, clusterName stri
+		}
+
+		if lbCert.Type == "lets_encrypt" {
+			l.ensureCertificateIDAnnot(service, lbCertID)
```
> I think it's not hurtful but it indicates that the design could be improved.

I don't think it's hurtful either, given that updating the service annotation is going to be a relatively rare edge case. If I remember correctly, Let's Encrypt certificates have a default lifetime of 3 months which is about how often we would end up updating the cert id annotation here.

I think additional effort to reduce the number of resource updates we make when handling this particular edge case would be more hurtful in terms of the opportunity cost of the time spent.

waynr commented on cloud-controller-manager/do/loadbalancers.go:
```diff
@@ -184,8 +197,19 @@ func newLoadBalancers(resources *resources, client *godo.Client, region string)
-func (l *loadBalancers) GetLoadBalancer(ctx context.Context, clusterName string, service *v1.Service) (*v1.LoadBalancerStatus, bool, error) {
-	lb, err := l.retrieveAndAnnotateLoadBalancer(ctx, service)
+func (l *loadBalancers) GetLoadBalancer(ctx context.Context, clusterName string, service *v1.Service) (lbs *v1.LoadBalancerStatus, truthy bool, err error) {
+	patcher := newServicePatcher(service)
```
@timoreimann @snormore here is an example of my solution for aggregating in-memory service modifications into a single PATCH call to the kubernetes api. The reason I chose to defer the Patch call is to reduce the chances of service changes happening that don't get captured before return. If you think that's too convoluted I'm fine identifying a more appropriate place to call `patcher.Patch`.

One thing that's worth noting is that in the defer methods as written if the function returns an error then we currently skip the patch. Do you think we should instead unconditionally always patch? I'm not certain myself which would be best.

timoreimann replied:
In general, I like the defer approach a lot. I'll comment around a few design details on more specific lines, but overall it seems like a good approach to ensure we include all patch-worthy annotations in a single invocation.

Yes, I think we should patch independently of whether we hit an error or not. That's what we did before, and the reason is this: even if our request to the LB API fails overall, the kind of patches we intend to do (LB ID / certificate updates) will help us avoid unnecessary follow-up API requests. For example, if we did not have the LB ID recorded in the Service yet and we only failed in a final LB update API request, persisting the LB ID annotation already would ensure that our API requests we executed in the controller-driven retry attempts could fetch the LB object efficiently in O(1) (vs. having to iterate through all LBs exhaustively in the worst case if the LB UUID was still lacking).

Similarly for the certificate ID where we could avoid reaching out to the Certificate API as soon as the certificate IDs matched.

The only thing we need to do then is to short-circuit our `Patch` call if the annotations match. That way, we don't even reach out to the API server if there's nothing worthwhile patching (and thus rule out a potential source of I/O erroring). Would appreciate if you could do the necessary adjustments. (We can copy [what upstream does](https://github.com/digitalocean/digitalocean-cloud-controller-manager/blob/67c685a86ca3b54897fa4f835c0e652844466a9a/vendor/k8s.io/kubernetes/pkg/controller/service/service_controller.go#L484-L486), which is to invoke `reflect.DeepCopy` on the annotations.)

waynr commented:
@timoreimann I've moved the service certificate check into a separate PR: https://github.com/digitalocean/digitalocean-cloud-controller-manager/pull/275

timoreimann commented on cloud-controller-manager/do/loadbalancers.go:
```diff
@@ -184,8 +197,19 @@ func newLoadBalancers(resources *resources, client *godo.Client, region string)
+		if err == nil {
+			err = patcher.Patch(l.resources.kclient, service)
+			if err != nil {
+				err = fmt.Errorf("failed to patch service: %s", err)
```
`patcher.Patch` returns annotated errors (that also include the Service name and namespace which this error currently doesn't), so we should return a bare `err` from that method call to avoid stuttering.

timoreimann replied:
If we go with patching regardless of the outer error (i.e., the value of `err` in line 203), we'll have to make sure that both an outer non-nil error and any non-nil error returned from `patcher.Patch` are propagated to the caller.

A convenient way to do that is to use [apimachinery's `Aggregate`](https://github.com/kubernetes/apimachinery/blob/release-1.15/pkg/util/errors/errors.go) helper: it basically aggregates (or wraps) multiple error values into a single error type that handles `nil` values appropriately and returns a single string representation of the errors. The usage in our case would look like this:

```go
import utilerrors "k8s.io/apimachinery/util/errors"
[...]

defer func() {
  perr := patcher.Patch(l.resources.kclient, service)
  errs := []errors{err, perr} // Errors can be nil; utilerrors.Aggregate will handle it correctly.
  err = utilerrors.NewAggregate(errs)
}
```

If you're curious, check out the [`Error()` implementation](https://github.com/kubernetes/apimachinery/blob/release-1.15/pkg/util/errors/errors.go#L64) to see what a string representation would look like.

timoreimann commented on cloud-controller-manager/do/loadbalancers.go:
```diff
@@ -171,6 +172,18 @@ type loadBalancers struct {
+}
+
+func (sp *servicePatcher) Patch(kclient kubernetes.Interface, updated *v1.Service) error {
+	return patchService(kclient, sp.base, updated)
```
`updated` is always pointing at the same `base` Service object we received prior to deep-copying it, right?

If that's the case, then we could also a field

```
updated *v1.Service
``` 

to `servicePatcher` and set it to `base` (again, prior to deep-copying), thereby saving on passing through `updated` here.

timoreimann replied:
Nit: Similarly, we could instantiate `servicePatcher` with a `kubernetes.Interface`.

timoreimann commented on cloud-controller-manager/do/loadbalancers.go:
```diff
@@ -275,27 +306,91 @@ func (l *loadBalancers) EnsureLoadBalancer(ctx context.Context, clusterName stri
+				err = fmt.Errorf("failed to patch service: %s", err)
+			}
+		}
+	}()
```
Since we have implemented the deferred method three times, how about attaching it to `servicePatcher` so that we can do a direct `defer.Patch(...)` call (and move the error checking / handling into the method too)?

timoreimann commented on cloud-controller-manager/do/loadbalancers.go:
```diff
@@ -360,20 +453,11 @@ func (l *loadBalancers) retrieveLoadBalancer(ctx context.Context, service *v1.Se
-	updated := service.DeepCopy()
-	if updated.ObjectMeta.Annotations == nil {
-		updated.ObjectMeta.Annotations = map[string]string{}
+func ensureServiceAnnotation(service *v1.Service, annotName, annotValue string) {
```
`ensure*` usually indicates that a piece of logic changes something or makes sure that the change is already applied. `ensureServiceAnnotation` changed to simply recording a service annotation, so I'd opt to rename it to something like `recordServiceAnnotation`.

timoreimann commented on cloud-controller-manager/do/certificates_test.go:
```diff
@@ -0,0 +1,318 @@
+			Type: certTypeLetsEncrypt,
+		}, newFakeOKResponse(), nil
+	}
+	lb, ok := f.store[certID]
```
copy-pasta: `lb` should be called `certificate` or `cert`.

timoreimann commented on cloud-controller-manager/do/certificates_test.go:
```diff
@@ -0,0 +1,318 @@
+				case "UpdateLoadBalancer":
+					err = lb.UpdateLoadBalancer(context.TODO(), "test", service, nodes)
+				default:
+					t.Errorf("unsupported loadbalancer method: %s", methodName)
```
Should be `t.Fatalf`.

timoreimann commented on cloud-controller-manager/do/certificates_test.go:
```diff
@@ -0,0 +1,318 @@
+
+type kvCertService struct {
+	store         map[string]*godo.Certificate
+	reflectorMode bool
```
We do not seem to set `reflectorMode` anywhere, can we remove it?

When in doubt (and FWIW), I generally prefer populating fakes with the right values over setting a special mode (except for simulating error cases, but that's not what we have here).

timoreimann commented on cloud-controller-manager/do/certificates_test.go:
```diff
@@ -0,0 +1,318 @@
+	}
+}
+
+func createServiceAndCert(lbID, certID, certType string) (*v1.Service, *godo.Certificate) {
```
refactor: I implemented a [service builder](https://github.com/digitalocean/digitalocean-cloud-controller-manager/blob/67c685a86ca3b54897fa4f835c0e652844466a9a/cloud-controller-manager/do/resources_test.go#L42) elsewhere which could be extended and leveraged for testing certificates.

For common service builder parameters, I also [created shortcuts at times](https://github.com/digitalocean/digitalocean-cloud-controller-manager/blob/67c685a86ca3b54897fa4f835c0e652844466a9a/cloud-controller-manager/do/resources_test.go#L94-L96).

timoreimann commented on cloud-controller-manager/do/loadbalancers.go:
```diff
@@ -275,27 +306,91 @@ func (l *loadBalancers) EnsureLoadBalancer(ctx context.Context, clusterName stri
+			id = rule.CertificateID
+			// CCM does not currently support multiple certificates on the
+			// loadbalancers it manages so we only need to grab the first one we find
+			break
```
refactor: we do not need the `id` variable: this line can return `rule.CertificateID` directly, and 319 can return an empty string.

timoreimann commented on cloud-controller-manager/do/loadbalancers.go:
```diff
@@ -275,27 +306,91 @@ func (l *loadBalancers) EnsureLoadBalancer(ctx context.Context, clusterName stri
+			if ok && respErr.Response.StatusCode == http.StatusNotFound {
+				return nil
+			}
+			return fmt.Errorf("failed to get DO certificate for lb: %s", err)
```
refactor: CCM should always spell out `load-balancer` in error messages.

timoreimann commented on cloud-controller-manager/do/loadbalancers.go:
```diff
@@ -275,27 +306,91 @@ func (l *loadBalancers) EnsureLoadBalancer(ctx context.Context, clusterName stri
+	return id
+}
+
+func (l *loadBalancers) checkAndUpdateLBAndServiceCerts(ctx context.Context, service *v1.Service, lbCertID, serviceCertID string) error {
```
Method name needs to be updated. Suggestion: `recordLetsEncryptLBCert`

(I've dropped the `check` prefix deliberately, don't feel it's necessary to encode the check-ness nature of code into names as we do those for short-circuiting all the times.)

timoreimann replied:
Can we add a GoDoc describing what the function does and in particular why we need it (i.e., explain the case of LE certs getting updated on LBs and CCM needing to sync with that)?

timoreimann commented on cloud-controller-manager/do/loadbalancers_test.go:
```diff
@@ -85,8 +86,38 @@ func (f *fakeLBService) RemoveForwardingRules(ctx context.Context, lbID string,
 	return f.removeForwardingRulesFn(ctx, lbID, rules...)
 }
 
+func newKVLBService(store map[string]*godo.LoadBalancer) fakeLBService {
```
Really appreciate moving towards an LB fake. 👍 

If this continues to work with the current code, then I'm all happy. FWIW, I'm working on a branch that [fakes the LB](https://github.com/digitalocean/digitalocean-cloud-controller-manager/blob/b1cb741066efa35a3ae3fdeedebfec6c350f208d/cloud-controller-manager/do/fakelbsvc_test.go) as well. It's a bit more sophisticated since I'm already covering more cases. Additionally, I found out that I had to do things like copying LB structs to ensure that modifying them in production code does not impact the fake store.

Feel free to take inspiration from my WIP branch or ignore it entirely as needed by your PR.

timoreimann commented on cloud-controller-manager/do/certificates_test.go:
```diff
@@ -0,0 +1,318 @@
+}
+
+func Test_LBaaSCertificateScenarios(t *testing.T) {
+	testcases := []struct {
```
refactor: `s/testcases/tests/` to be consistent with the naming style we employ on more recent tests.

timoreimann commented on cloud-controller-manager/do/certificates_test.go:
```diff
@@ -0,0 +1,318 @@
+		},
+	}
+
+	for _, tc := range testcases {
```
refactor: `s/tc/test/` to be consistent with the naming style we employ or recent tests.

timoreimann commented on cloud-controller-manager/do/certificates_test.go:
```diff
@@ -0,0 +1,318 @@
+					t.Logf("actual: %s", serviceCertID)
+				}
+
+				godoLoadBalancer, _, err := lbService.Get(context.TODO(), getLoadBalancerID(service))
```
This should use `context.Background()` instead of `context.TODO()` because the latter is supposed to be used

>when it's unclear which Context to use or it is not yet available (because the surrounding function has not yet been extended to accept a Context parameter)

([source](https://golang.org/pkg/context/#TODO))

timoreimann commented on cloud-controller-manager/do/certificates_test.go:
```diff
@@ -0,0 +1,318 @@
+				}
+				lbCertID := getCertificateIDFromLB(godoLoadBalancer)
+				if tc.expectedLBCertID != lbCertID {
+					t.Error("unexpected loadbalancer certificate ID")
```
refactor: I know many of our existing tests follow this three line pattern, which is overly verbose -- especially for assertions on primitive types.

```
t.Errorf("got certificate ID %s, want %s", lbCertID, tc.expectedLBCertID)
```

should be a win on all fronts.

timoreimann commented on cloud-controller-manager/do/certificates_test.go:
```diff
@@ -0,0 +1,318 @@
+					t.Fatalf("failed to get loadbalancer %q from fake client: %s", getLoadBalancerID(service), err)
+				}
+				lbCertID := getCertificateIDFromLB(godoLoadBalancer)
+				if tc.expectedLBCertID != lbCertID {
```
refactor: [revert comparison order](https://github.com/golang/go/wiki/CodeReviewComments#useful-test-failures).

timoreimann commented on cloud-controller-manager/do/certificates_test.go:
```diff
@@ -0,0 +1,318 @@
+
+		// lets_encrypt test cases
+		{
+			name: "[letsencrypt] LB cert ID and service cert ID match ",
```
Trailing blank at the end of the string.

timoreimann commented on cloud-controller-manager/do/loadbalancers_test.go:
```diff
@@ -100,6 +131,31 @@ func createLB() *godo.LoadBalancer {
 	}
 }
 
+func createHTTPSLB(entryPort, targetPort int, lbID, certID, certType string) (*godo.LoadBalancer, *godo.Certificate) {
```
Should probably have a better name like `createHTTPSLBAndCertificate`.

timoreimann replied:
refactor: What I'd prefer though is if the function does one thing only (creation of LBs), with the creation of the certificate happening outside and the certificate ID being passed into this function (i.e., follow the S in SOLID).

waynr replied:
I DRY'd up this up by removing the `entryPort` and `targetPort` parameters but did not refactor.

timoreimann commented on cloud-controller-manager/do/certificates_test.go:
```diff
@@ -0,0 +1,318 @@
+		err                   error
+	}{
+		{
+			name: "default test values, tls not enabled",
```
What is the distinct purpose of this test? Isn't it covered by the other tests below?

waynr replied:
it was originally intended to validate the default LB and Cert "fake" services when I was attempting to follow the example of overriding interface functions on a per test case level. I don't think it's necessary anymore now that I've implemented the key-value store backed fakes so I'll remove it.

timoreimann commented on cloud-controller-manager/do/certificates_test.go:
```diff
@@ -0,0 +1,318 @@
+		{
+			name: "[letsencrypt] LB cert ID and service cert ID differ and both certs exist",
+			setupFn: func(lbService fakeLBService, certService kvCertService) *v1.Service {
+				lb, cert := createHTTPSLB(443, 30000, "test-lb-id", "test-cert-id", certTypeLetsEncrypt)
```
nit: would name `test-cert-id` as `lb-cert-id` as "test" has fairly little meaning.

waynr replied:
Ah yeah, I meant to do this eventually but forgot -- I was originally following established naming conventions I saw in the loadbalancer tests.

waynr commented:
@timoreimann I've marked conversations where I addressed your comments as resolved FYI.

@snormore @timoreimann 
Conformance tests just passed for me against clusters with the latest head of this branch deployed. I have manually validated that none of the changes here affect creation or deletion of Services. I have also validated the new behavior for Let's Encrypt certs with the following workflow:

* Create a DO cluster using a testing version of DOKS that includes the changes in this PR
* Create a cert on the command line:
```
doctl compute certificate create --type lets_encrypt --dns-names testing.waynr.net --name testing.waynr.net
```
* Create a `Service` object on the cluster with `service.beta.kubernetes.io/do-loadbalancer-certificate-id` set to the UUId of the aforementioned certificate
* Observe the creation of a load-balancer in my DO cloud account
* Observe the creation of a `testing.waynr.net` subdomain in my DNS settings
* Create another cert on the command line:
```
doctl compute certificate create --type lets_encrypt --dns-names testing2.waynr.net --name testing2.waynr.net
```
* Modify the load-balancer's forwarding rules to use the new `testing2.waynr.net` cert.
* Add a new node pool to the cluster through the cloud UI
* Observe the `Service` annotations updated with the new `testing2.waynr.net` cert UUID
